develop/data-generation/block-models.md

```diff
@@ -0,0 +1,216 @@
+
+@[code lang=java transcludeWith=:::datagen-model:door-and-trapdoor](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+Doors and trapdoors are a little different. Here, you have to make three new textures - two for the door, and one for the trapdoor.
+
+1. The door:
+    - It has two parts - the upper half and the lower half. **Each needs its own texture:** in this case `ruby_door_top` for the upper half and `ruby_door_bottom` for the lower.
+    - The `registerDoor()` method will create models for all orientations of the door, both open and closed.
+    - **You also need an item texture!** Put it in `assets/<mod_id>/textures/item/` folder.
+2. The trapdoor:
+   - Here, you need only one texture, in this case named `ruby_trapdoor`. It will be used for all sides.
+   - Since `TrapdoorBlock` has a `FACING` property, you can use the commented out method to generate model files with rotated textures = the trapdoor will be "orientable". Otherwise, it will look the same no matter the direction it's facing.
+
+<DownloadEntry visualURL="/assets/develop/data-generation/block-model/ruby_door_trapdoor_big.png" downloadURL="/assets/develop/data-generation/block-model/ruby_door_trapdoor_textures.zip">Ruby Door and Trapdoor</DownloadEntry>
+
+## Custom Block Models {#custom-block-models}
+
+In this section, we'll create the models for a Vertical Oak Log Slab, with Oak Log textures.
+
+_Points 2. - 6. are declared in an inner static helper class called `CustomBlockStateModelGenerator`._
+
+### Custom Block Class {#custom-block-class}
+
+Create a `VerticalSlab` block with a `FACING` property and a `SINGLE` boolean property, like in the [Block States](../blocks/blockstates) tutorial. `SINGLE` will indicate if there are both slabs.
+Then you should override `getOutlineShape` and `getCollisionShape`, so that the outline is rendered correctly, and the block has the correct collision shape.
+
+@[code lang=java transcludeWith=:::datagen-model-custom:voxels](@/reference/latest/src/main/java/com/example/docs/block/custom/VerticalSlabBlock.java)
+
+@[code lang=java transcludeWith=:::datagen-model-custom:collision](@/reference/latest/src/main/java/com/example/docs/block/custom/VerticalSlabBlock.java)
+
+Also override the `canReplace()` method, otherwise you couldn't make the slab a full block.
+
+@[code lang=java transcludeWith=:::datagen-model-custom:replace](@/reference/latest/src/main/java/com/example/docs/block/custom/VerticalSlabBlock.java)
+
+And you're done! You can now test the block out and place it in game.
+
+### Parent Block Model {#parent-block-model}
+
+Now, let's create a parent block model. It will determine the size, position in hand or other slots and the `x` and `y` coordinates of the texture.
+I highly recommend using [Blockbench](https://www.blockbench.net/) for this, as making it manually is a really tedious process. It should look something like this:
```

Comment:
Shouldnt be "I recommend", should be "It's recommended to use an editor such as Blockbench ..."